wp-seo-main.php

```diff
@@ -374,6 +382,46 @@ function wpseo_init() {
 	}
 }
 
+/**
+ * Detects if we can invalidate the opcache and does so.
+ *
+ * @return bool
+ */
+function wpseo_maybe_invalidate_opcache() {
+	require_once ABSPATH . 'wp-admin/includes/file.php';
+
+	if ( ! function_exists( 'wp_opcache_invalidate' ) || ! function_exists( 'wp_opcache_invalidate_directory' ) ) {
+		return false;
+	}
+
+	$access_type = get_filesystem_method();
+	if ( $access_type !== 'direct' ) {
+		return false;
+	}
+
+	// Prevent any output requesting credentials, even though we should be safe in 'direct' mode.
+	ob_start();
+	$credentials = request_filesystem_credentials( admin_url() );
+	ob_end_clean();
+
+	if ( $credentials === false || ! WP_Filesystem( $credentials ) ) {
+		return false;
+	}
+
+	// Since WP 5.5.
+	wp_opcache_invalidate( WPSEO_PATH . 'wp-seo-main.php' );
+	wp_opcache_invalidate( WPSEO_PATH . 'wp-seo.php' );
+	// Since WP 6.2.
+	wp_opcache_invalidate_directory( WPSEO_PATH . 'admin' );
```

Comment:
The `wp_opcache_invalidate_directory()` function is extremely wasteful:

- It recursively iterates over _all_ files in the provided directories, not only the PHP files.
- For every single file, it first does multiple environment checks with some heavy computation (like `ini_get()`, `realpath()`).
- And _only then_ it checks whether the provided file is even a PHP file to begin with.

I'll open a ticket for this in Core, but in the meantime I'm wondering if we shouldn't prefer to skip the use of `wp_opcache_invalidate_directory()` and rather provide our own iteration code instead which would be much faster and less wasteful.